### DISCUSSION TITLE
AWS Events: Create Event Bridge target from an rule defined in a different stack

### DISCUSSION THREAD

**Discussion by lucashfreitas** [aws/aws-cdk#21997] [Q&A]
I have a CDK stack that defines all event bridges `rules` and exports their ARNs as outputs. I would like to create a target for that rules in other stacks where my consumers are. 

I can access the role from an exported ARN without issues:

``` typescript
    //that works
    const eventBusRule = cdk.aws_events.Rule.fromEventRuleArn(
      this,
      "EventBridgeRuleArn",
      'exportedRuleArn
    );
```

The Object returned from `FromEventRuleArn` doesn't have properties to AddTarget. `aws_events` also doesn't have any way to create a new target passing a rule as a reference/option. 

<img width="709" alt="image" src="https://user-images.githubusercontent.com/11416392/189576123-25cfb5fc-a7c1-4385-86c4-0148a7defc90.png">

**Comment by pahud** [comment]
`fromEventRuleArn` actually returns [IRule](https://github.com/aws/aws-cdk/blob/cea1039e3664fdfa89c6f00cdaeb1a0185a12678/packages/%40aws-cdk/aws-events/lib/rule-ref.ts#L6)
https://github.com/aws/aws-cdk/blob/cea1039e3664fdfa89c6f00cdaeb1a0185a12678/packages/%40aws-cdk/aws-events/lib/rule.ts#L100

If you need the [addTarget()](https://github.com/aws/aws-cdk/blob/cea1039e3664fdfa89c6f00cdaeb1a0185a12678/packages/%40aws-cdk/aws-events/lib/rule.ts#L168) method you will need to create a new [Rule](https://github.com/aws/aws-cdk/blob/cea1039e3664fdfa89c6f00cdaeb1a0185a12678/packages/%40aws-cdk/aws-events/lib/rule.ts#L91) object

**Reply by lucashfreitas** [reply]
It is possible to create a new rule object from an existent resource Without creating a new one?

**Reply by pahud** [reply]
When you import an existing `rule`, which is not created and managed by CDK, you actually get an `IRule` which is sort of a reference you can only refer to its attribute like Arn or ID. In this case you actually didn't instantiate the `Rule` L2 construct class.  In CDK when you "import" any existing resource, you will not get a full L2 construct class of that resource as it's not "created and managed" by CDK.

**Reply by lucashfreitas** [reply]
Yeah, I am aware of this, trying to find out if there is any workaround that can be used to create a Target Object with a given rule ARN. Will have a look on Cloudformation docs :) Thank you very much @pahud.

**Reply by pahud** [reply]
@lucashfreitas If we look at [AWS::Events::Rule](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-events-rule.html) there's a `Targets` property where you can define multiple events targets for the rule to be created by cloudformation:

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-events-rule.html#cfn-events-rule-targets

Given the rule is already created in your case and all we know is just its ARN, I don't think cloudformation is able to "link" the target(s) to the existing rule. I believe that's the key limitation.

**Reply by pahud** [reply]
Looking at [putTargets](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EventBridge.html#putTargets-property) method in SDK, I believe it should be possible to create a custom construct with [custom-resources](https://github.com/aws/aws-cdk/tree/main/packages/%40aws-cdk/custom-resources) and pass the `Rule` and `EventBusName` as the props to this custom construct and it should technically possible to add "target(s)" to the existing rule with CDK. Under the hood, it just calls the `putTargets()` to add target(s) to the existing rule.
